modules/mbedtls/zephyr_init.c

```diff
@@ -40,12 +46,56 @@ static void init_heap(void)
 #define init_heap(...)
 #endif /* CONFIG_MBEDTLS_ENABLE_HEAP && MBEDTLS_MEMORY_BUFFER_ALLOC_C */
 
+static const struct device *entropy_dev =
+			DEVICE_DT_GET_OR_NULL(DT_CHOSEN(zephyr_entropy));
+
+int mbedtls_hardware_poll(void *data, unsigned char *output, size_t len,
+			  size_t *olen)
+{
+	int ret;
+	uint16_t request_len = len > UINT16_MAX ? UINT16_MAX : len;
+
+	ARG_UNUSED(data);
+
+	if (output == NULL || olen == NULL || len == 0) {
+		return MBEDTLS_ERR_ENTROPY_SOURCE_FAILED;
+	}
+
+	if (!IS_ENABLED(CONFIG_ENTROPY_HAS_DRIVER)) {
+		sys_rand_get(output, len);
+		*olen = len;
+
+		return 0;
+	}
+
+	__ASSERT(entropy_dev != NULL, "No entropy device found");
+
+	if (!device_is_ready(entropy_dev)) {
+		return MBEDTLS_ERR_ENTROPY_SOURCE_FAILED;
+	}
+
+	ret = entropy_get_entropy(entropy_dev, (uint8_t *)output, request_len);
+	if (ret < 0) {
+		return MBEDTLS_ERR_ENTROPY_SOURCE_FAILED;
+	}
+
+	*olen = request_len;
+
+	return 0;
+}
+
 static int _mbedtls_init(const struct device *device)
 {
 	ARG_UNUSED(device);
 
 	init_heap();
 
+	if (IS_ENABLED(CONFIG_MBEDTLS_ZEPHYR_ENTROPY) &&
+	    !IS_ENABLED(CONFIG_ENTROPY_HAS_DRIVER)) {
+		LOG_WRN("No entropy device on the system, "
+			"using fake entropy source!");
+	}
+
```

Comment:
Since that is dependent on build configuration only, can we issue this warning from cmake at the project configuration / build phase?

Reply:
Yes, moved the warning to the CMakeLists.txt instead